plonk/src/lagrange.rs

```diff
@@ -0,0 +1,224 @@
+//! Utilities for Lagrange interpolations, evaluations, coefficients for a
+//! polynomial
+
+use ark_ff::FftField;
+use ark_poly::{EvaluationDomain, Radix2EvaluationDomain};
+use ark_std::{ops::Range, vec, vec::Vec};
+
+// TODO: (alex) include these APIs upstream in arkworks directly
+
+/// A helper trait for computing Lagrange coefficients of an evaluation domain
+///
+/// from arkworks:
+/// Evaluate all Lagrange polynomials at tau to get the lagrange coefficients.
+/// Define the following as
+/// - H: The coset we are in, with generator g and offset h
+/// - n: The size of the coset H
+/// - Z_H: The vanishing polynomial for H. Z_H(x) = prod_{i in n} (x - hg^i) =
+///   x^n - h^n
+/// - v_i: A sequence of values, where v_0 = 1/(n * h^(n-1)), and v_{i + 1} = g
+///   * v_i
+///
+/// We then compute L_{i,H}(tau) as `L_{i,H}(tau) = Z_H(tau) * v_i / (tau - h *
+/// g^i)`
+#[allow(dead_code)]
+pub(crate) trait LagrangeCoeffs<F: FftField> {
+    /// Returns the first coefficient: `L_{0, Domain}(tau)`
+    fn first_lagrange_coeff(&self, tau: F) -> F;
+    /// Returns the last coefficient: `L_{n-1, Domain}(tau)`
+    fn last_lagrange_coeff(&self, tau: F) -> F;
+    /// Return a list of coefficients for `L_{range, Domain}(tau)`
+    fn lagrange_coeffs_for_range(&self, range: Range<usize>, tau: F) -> Vec<F>;
+}
+
+impl<F: FftField> LagrangeCoeffs<F> for Radix2EvaluationDomain<F> {
+    // L_0(tau) = Z_H(tau) * g^0 / (n * h^(n-1) *(tau - h * g^0))
+    // with g^0 = 1
+    // special care when tau in H, as both numerator and denominator is zero
+    fn first_lagrange_coeff(&self, tau: F) -> F {
+        let offset = self.coset_offset();
+        let z_h_at_tau = self.evaluate_vanishing_polynomial(tau);
+        if z_h_at_tau.is_zero() {
+            if tau == F::one() * offset {
```

Comment:
How about putting this check before computing z_h_at_tau? Then we can save a eval_vanish_poly ops when tau = h. And for other tau \in H, we can now directly return zero without if-else condition.

Reply:
> How about putting this check before computing z_h_at_tau? Then we can save a eval_vanish_poly ops when tau = h

sure, I'll do that.

>  And for other tau \in H, we can now directly return zero without if-else condition.

do you suggest we loop over all elements in the group and find matches? the easiest would still be evaluating the vanishing poly (which would still be a if-else)

Reply:
done in [2249534](https://github.com/EspressoSystems/jellyfish/pull/611/commits/224953442d24b72679f615db5b3dbb66aa1f72db)